### PR TITLE
Fix reinplace alias checks for index_put

### LINKED ISSUE #156786: [inductor] [cpu] `torch.nn.Embedding-torch.index_copy` outputs inconsistent results on cpu inductor
### 🐛 Describe the bug

**symptom**: `torch.nn.Embedding-torch.index_copy` outputs inconsistent results on cpu inductor. Note that graph break can eliminate errors. I guess some optimization between `Embedding` and `index_copy` is processed incorrectly on CPP backend
**codegen backend**: only CPP
```python
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch._inductor import config

config.fallback_random = True
torch.set_grad_enabled(False)
torch.manual_seed(0)


class Model(torch.nn.Module):

    def __init__(self):
        super().__init__()
        self.embedding = torch.nn.Embedding(num_embeddings=10, embedding_dim=4)

    def forward(self, x, indices):
        x_embed = self.embedding(x)
        # torch._dynamo.graph_break()  # graph break can eliminate errors
        out = torch.index_copy(x_embed, dim=1, index=indices, source=x_embed)
        return out


model = Model()


x = torch.randint(0, 10, (4,))
indices = torch.tensor([3, 1, 0, 2])

inputs = [x, indices]


def run_test(model, inputs, device, backend):
    torch.manual_seed(0)
    model = model.to(device)
    inputs = [x.to(device) for x in inputs]
    if backend != "eager":
        model = torch.compile(model, backend=backend)
    torch.manual_seed(0)
    output = model(*inputs)
    return output


device = 'cpu'
output = run_test(model, inputs, device, 'eager')
c_output = run_test(model, inputs, device, 'inductor')
print(torch.allclose(output, c_output, rtol=1e-3, atol=1e-3))
print(torch.max(torch.abs(c_output - output)))
```

### Error logs

CPP
```
False
tensor(0.6628)
```
triton
```
True
tensor(0., device='cuda:0')
```

### Versions

2.8.0.dev20250623+cu126

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185450

Inductor's reinplace pass can rewrite functional index_put into its in-place variant when the destination has no later observable uses. That check was only reasoning about later uses of the destination tensor. It did not account for operands that index_put reads during the operation.

When self aliases either the values tensor or a tensor entry in the indices list, rewriting aten.index_put.default to aten.index_put_.default changes a functional update into an in-place read/write overlap. That is unsafe: eager either needs functional clone semantics or raises the usual overlap error for the in-place form. The same hazard applies to the _unsafe_index_put mapping, and the extra_check hook also has to handle with_effects nodes where inner op arguments are shifted by the token and op operands.

This adds an index_put-specific reinplace eligibility check that rejects overlap between self and all read tensor operands: values plus every tensor nested in indices. The check is shared by aten.index_put.default and aten._unsafe_index_put.default, and it accounts for with_effects positional argument offsets.

An older abandoned draft PR for #156786 only checked the last positional argument, which misses schemas with an accumulate argument and does not cover indices. This version keys off the operator schema positions instead.

Fixes #156786
Generated by my agent

Test Plan:
- python -m pytest test/inductor/test_inplacing_pass.py -k 'index_put_source_alias_should_not_reinplace or index_put_indices_alias_should_not_reinplace or unsafe_index_put_source_alias_should_not_reinplace' -q
- python -m pytest test/inductor/test_inplacing_pass.py -q
- lintrunner -a

Benchmark Results:
Microbenchmark measured reinplace_inplaceable_ops_core on a non-aliasing index_put FX graph, using 7 samples of 3000 pass runs after warmup.

main:
- samples_us_per_pass= [443.35, 394.77, 411.03, 398.33, 398.33, 458.74, 404.95]
- median_us_per_pass= 404.95

fix:
- samples_us_per_pass= [429.71, 412.25, 414.45, 415.83, 417.07, 489.16, 418.41]
- median_us_per_pass= 417.07

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says the only failing check was the unrelated linux-jammy-py3.14-clang18 serialization crossref shard, where test_serialization_2gb_file was repeatedly SIGKILLed. I added the missing release notes: inductor label and reran the failed CI jobs so Dr. CI can reclassify/update.

File: torch/_inductor/fx_passes/reinplace.py [hunk 1]
```diff
@@ -370,9 +370,54 @@ def can_fuse():
...
+    if not isinstance(lhs, torch.fx.Node) or not isinstance(rhs, torch.fx.Node):
+        return False
+
+    lhs_storage = get_node_storage(lhs)
+    rhs_storage = get_node_storage(rhs)
+    if lhs_storage is None or lhs_storage != rhs_storage:
+        return False
+
+    try:
+        return len(compute_overlapping_tensors([lhs.meta["val"], rhs.meta["val"]])) != 0
```
Comment (mlazos) [inline torch/_inductor/fx_passes/reinplace.py:383]: _Reviewed by me, this may be an issue._

_AI-generated analysis (Codex):_

> **Cross-dtype storage overlap is missed**
> 
> `compute_overlapping_tensors` compares dtype-relative element offsets. An int32 `self` at offset 4 and an int64 index view at offset 2 share the same starting byte but are reported as disjoint. The pass then rewrites functional `index_put`, which succeeds, to `index_put_`, which raises an overlap RuntimeError.
